codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/SemanticVersion.java

```diff
@@ -0,0 +1,327 @@
+        patch = builder.patch;
+        extra = builder.extra;
+    }
+
+    /**
+     * The semantic version prefix present before the major version.
+     *
+     * @return the optional prefix
+     */
+    public Optional<String> getPrefix() {
+        return Optional.ofNullable(prefix);
+    }
+
+    /**
+     * The major version number.
+     *
+     * @return the major version
+     */
+    public int getMajor() {
+        return major;
+    }
+
+    /**
+     * The minor version number.
+     *
+     * @return the minor version
+     */
+    public int getMinor() {
+        return minor;
+    }
+
+    /**
+     * The patch version number.
+     *
+     * @return the patch version
+     */
+    public int getPatch() {
+        return patch;
+    }
+
+    /**
+     * The optional extra string containing the pre-release and build information.
+     *
+     * @return the optional extra string
+     */
+    public Optional<String> getExtra() {
+        return Optional.ofNullable(extra);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder();
+
+        if (getPrefix().isPresent()) {
+            builder.append(getPrefix().get());
+        }
+
+        builder.append(getMajor());
+        builder.append('.');
+        builder.append(getMinor());
+        builder.append('.');
+        builder.append(getPatch());
+        if (getExtra().isPresent()) {
+            builder.append(getExtra().get());
+        }
+
+        return builder.toString();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        SemanticVersion that = (SemanticVersion) o;
+        return getMajor() == that.getMajor()
+                && getMinor() == that.getMinor()
+                && getPatch() == that.getPatch()
+                && getPrefix().equals(that.getPrefix())
+                && getExtra().equals(that.getExtra());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getPrefix(), getMajor(), getMinor(), getPatch(), getExtra());
+    }
+
+    /**
+     * Parse a semantic version string into a {@link SemanticVersion}.
+     *
+     * @param version the semantic version string
+     * @return the SemanticVersion representing the parsed value
+     */
+    public static SemanticVersion parseVersion(String version) {
+        String[] components = version.split("\\.", 3);
+
+        if (components.length != 3) {
+            throw newInvalidSemanticVersion();
+        }
+
+        char[] parseArr = components[0].toCharArray();
+        int position = 0;
+
+        StringBuilder prefixBuilder = new StringBuilder();
+        while (position < parseArr.length && !Character.isDigit(parseArr[position])) {
+            prefixBuilder.append(parseArr[position]);
+            position++;
+        }
```

Comment:
This seems to assume first component will follow `(a-zA-Z)*[0-9]*`, if so, may be we should break the loop when we find the first character that is a digit.

Reply:
Prefixes are optional, (this is just to support Go's prefixing with `v`), and I don't see why any sort of break here is required. This will only loop while the current position is not a digit. So this will effectively strip any leading text on the major version. Test cases cover this.

Reply:
Correct, I read the while condition incorrectly. This seems to do the right thing.